Add CSP check for inline style attribute

TimvdLippe opened the pull request:
To be able to abort the update, extract the functionality into a separate method. Otherwise, we don't run the `node.rev_version` at the end, which according to the comment is probably important.

Not all `style-src` tests pass and I don't fully understand why yet, but I presume it has to do with some special quirks of stylesheets that other CSP checks don't have. All `style-src-attr-elem` tests pass though.

Part of #4577

TimvdLippe committed:
Add CSP check for inline style attribute

To be able to abort the update, extract the functionality into a
separate method. Otherwise, we don't run the `node.rev_version`
at the end, which according to the comment is probably important.

Not all `style-src` tests pass and I don't fully understand why yet,
but I presume it has to do with some special quirks of stylesheets
that other CSP checks don't have. All `style-src-attr-elem` tests
pass though.

Part of #4577

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/14914232749

jdm reviewed (approved):
Nice and simple.

Edit: Woah, those CI results don't look great.

jdm commented on components/script/dom/element.rs:
```diff
@@ -2120,6 +2121,59 @@ impl Element {
+        if doc.should_elements_inline_type_behavior_be_blocked(
+            self,
+            csp::InlineCheckType::StyleAttribute,
+            source,
```
This source value is incorrect. It gets compared against the policy source values like `'self'`, so every inline style change is blocked by CSP checks right now.

jdm replied:
Sorry, I did some more reading and realized that I was misunderstanding what `style-src 'self'` is supposed to do, and that the spec text describes exactly what the code is doing.

jdm commented:
I'm having trouble figuring out why Firefox and Safari allow setting this style attribute but Servo doesn't:
```html
<meta http-equiv="Content-Security-Policy" content="style-src 'self'">
<div>aaaaaaa</div>
<script>
  let e = document.querySelector('div');
  e.style.background = "yellow";
  console.log(e.style.background);
</script>
```

jdm commented:
Huh, other browsers apparently do this:
```
Also no CSP checks are done whenever setting a style property:
node.style.color = "green"; will always pass

but setting the value of the style attribute does trigger CSP checks:
node.style = "color: green;"; goes through the CSP inline checks
```
from https://github.com/w3c/webappsec-csp/issues/212#issuecomment-445263798

jdm commented:
It may be enough to gate the two calls to parse_style_attribute in component/script with the CSP checks.

TimvdLippe commented:
Thanks for the investigation and detailed explanation! I am surprised that I was able to crash Servo with a borrow issue, as I thought that's what safety Rust provides with the borrow checker? Or is that because `&**` does a lot of nasty heavy-lifting?

Anyways, I moved the check prior to the `parse_style_attribute` now and the tests locally are fine. Let's see what CI says: https://github.com/TimvdLippe/servo/actions/runs/14933782551

Also, I didn't add the check for `cssText` (the other usage of `parse_style_attribute`), since that shouldn't be blocked. This wasn't covered with a test yet, so I added a new one and confirmed that Chrome and Firefox pass on it.

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#52444) with upstreamable changes.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52444).

jdm commented:
> Thanks for the investigation and detailed explanation! I am surprised that I was able to crash Servo with a borrow issue, as I thought that's what safety Rust provides with the borrow checker? Or is that because `&**` does a lot of nasty heavy-lifting?
 
This is where precise language is important. The test harness calls it a crash because the browser process disappears before the test was complete. Servo panicked due to a dynamic borrow failure—the DOM implementation is full of shared ownership that doesn't fit into Rust's static borrow checker, so we use types like RefCell to enforce memory safety at runtime. In this case it detected an attempt to perform two mutable borrows simultaneously on the same piece of memory and panicked (safely) rather than crashing (unsafely (eg. a segfault)) or allowing undefined behavior.

The end result may look the same to a user (the browser stops working) but panics don't turn into memory safety vulnerabilities.
